File: backend/app/protocol.py
from __future__ import annotations

import json
import struct
import zlib
from dataclasses import dataclass
from enum import IntEnum
from typing import Any

MAGIC = b"IGNS"
VERSION = 1
HEADER = struct.Struct("!4sBBHIQQI")
HEADER_SIZE = HEADER.size


class ProtocolError(ValueError):
    pass


class PacketType(IntEnum):
    HELLO = 1
    HELLO_ACK = 2
    FRAME = 10
    DETECTIONS = 11
    INCIDENT_UPDATE = 12
    INCIDENT_TIMELINE_EVENT = 13
    HEALTH_UPDATE = 14
    EVIDENCE_MANIFEST = 15
    LOG_EVENT = 16
    CONFIG_ACK = 20
    CONFIG_UPDATE = 100
    OCCUPANT_CONFIRM = 101
    OCCUPANT_CANCEL = 102
    MANUAL_RESET = 103
    CALL_STATUS_UPDATE = 104
    PING = 105
    PONG = 106


@dataclass(frozen=True, slots=True)
class Packet:
    packet_type: PacketType
    flags: int
    sequence: int
    monotonic_ns: int
    payload: bytes

    def json(self) -> dict[str, Any]:
        try:
            value = json.loads(self.payload.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise ProtocolError("invalid JSON payload") from exc
        if not isinstance(value, dict):
            raise ProtocolError("JSON payload must be an object")
        return value


def encode_packet(
    packet_type: PacketType,
    payload: bytes | dict[str, Any],
    sequence: int,
    monotonic_ns: int,
    flags: int = 0,
) -> bytes:
    if isinstance(payload, dict):
        payload = json.dumps(payload, separators=(",", ":"), ensure_ascii=True).encode("utf-8")
    crc = zlib.crc32(payload) & 0xFFFFFFFF
    return HEADER.pack(
        MAGIC,
        VERSION,
        int(packet_type),
        flags,
        len(payload),
        sequence,
        monotonic_ns,
        crc,
    ) + payload


class PacketParser:
    def __init__(self, max_payload_bytes: int = 2_097_152):
        self.max_payload_bytes = max_payload_bytes
        self._buffer = bytearray()
        self._expected: tuple[PacketType, int, int, int, int, int] | None = None

    @property
    def buffered_bytes(self) -> int:
        return len(self._buffer)

    def feed(self, chunk: bytes) -> list[Packet]:
        if len(self._buffer) + len(chunk) > self.max_payload_bytes + HEADER_SIZE:
            raise ProtocolError("connection buffer exceeded maximum packet size")
        self._buffer.extend(chunk)
        packets: list[Packet] = []
        while True:
            if self._expected is None:
                if len(self._buffer) < HEADER_SIZE:
                    break
                raw_header = bytes(self._buffer[:HEADER_SIZE])
                del self._buffer[:HEADER_SIZE]
                magic, version, raw_type, flags, length, sequence, monotonic_ns, crc = (
                    HEADER.unpack(raw_header)
                )
                if magic != MAGIC:
                    raise ProtocolError("invalid packet magic")
                if version != VERSION:
                    raise ProtocolError("unsupported protocol version")
                try:
                    packet_type = PacketType(raw_type)
                except ValueError as exc:
                    raise ProtocolError("unknown packet type") from exc
                if length > self.max_payload_bytes:
                    raise ProtocolError("payload exceeds configured maximum")
                self._expected = (packet_type, flags, length, sequence, monotonic_ns, crc)
            packet_type, flags, length, sequence, monotonic_ns, crc = self._expected
            if len(self._buffer) < length:
                break
            payload = bytes(self._buffer[:length])
            del self._buffer[:length]
            self._expected = None
            if zlib.crc32(payload) & 0xFFFFFFFF != crc:
                raise ProtocolError("payload CRC mismatch")
            packets.append(Packet(packet_type, flags, sequence, monotonic_ns, payload))
        return packets


def encode_frame_payload(metadata: dict[str, Any], jpeg: bytes) -> bytes:
    encoded = json.dumps(metadata, separators=(",", ":"), ensure_ascii=True).encode("utf-8")
    return struct.pack("!I", len(encoded)) + encoded + jpeg


def decode_frame_payload(payload: bytes) -> tuple[dict[str, Any], bytes]:
    if len(payload) < 4:
        raise ProtocolError("truncated frame metadata length")
    (length,) = struct.unpack("!I", payload[:4])
    if length == 0 or 4 + length >= len(payload):
        raise ProtocolError("invalid frame metadata length")
    try:
        metadata = json.loads(payload[4 : 4 + length].decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ProtocolError("invalid frame metadata JSON") from exc
    jpeg = payload[4 + length :]
    if not isinstance(metadata, dict):
        raise ProtocolError("frame metadata must be an object")
    if len(jpeg) < 4 or not jpeg.startswith(b"\xff\xd8") or not jpeg.endswith(b"\xff\xd9"):
        raise ProtocolError("truncated or invalid JPEG")
    return metadata, jpeg

File: qnx/common/src/protocol.cpp
#include "ignis/protocol.hpp"

#include <algorithm>
#include <array>
#include <cstring>

namespace ignis {

namespace {

void write_u16(std::uint8_t* output, std::uint16_t value) {
    output[0] = static_cast<std::uint8_t>(value >> 8);
    output[1] = static_cast<std::uint8_t>(value);
}

void write_u32(std::uint8_t* output, std::uint32_t value) {
    for (int index = 0; index < 4; ++index)
        output[index] = static_cast<std::uint8_t>(value >> (24 - 8 * index));
}

void write_u64(std::uint8_t* output, std::uint64_t value) {
    for (int index = 0; index < 8; ++index)
        output[index] = static_cast<std::uint8_t>(value >> (56 - 8 * index));
}

std::uint16_t read_u16(const std::uint8_t* input) {
    return static_cast<std::uint16_t>((input[0] << 8) | input[1]);
}

std::uint32_t read_u32(const std::uint8_t* input) {
    std::uint32_t value = 0;
    for (int index = 0; index < 4; ++index) value = (value << 8) | input[index];
    return value;
}

std::uint64_t read_u64(const std::uint8_t* input) {
    std::uint64_t value = 0;
    for (int index = 0; index < 8; ++index) value = (value << 8) | input[index];
    return value;
}

}  // namespace

bool known_packet_type(std::uint8_t value) noexcept {
    switch (static_cast<PacketType>(value)) {
        case PacketType::Hello:
        case PacketType::HelloAck:
        case PacketType::Frame:
        case PacketType::Detections:
        case PacketType::IncidentUpdate:
        case PacketType::IncidentTimelineEvent:
        case PacketType::HealthUpdate:
        case PacketType::EvidenceManifest:
        case PacketType::LogEvent:
        case PacketType::ConfigAck:
        case PacketType::ConfigUpdate:
        case PacketType::OccupantConfirm:
        case PacketType::OccupantCancel:
        case PacketType::ManualReset:
        case PacketType::CallStatusUpdate:
        case PacketType::Ping:
        case PacketType::Pong: return true;
    }
    return false;
}

std::uint32_t crc32(const std::uint8_t* data, std::size_t size) noexcept {
    std::uint32_t crc = 0xFFFFFFFFU;
    for (std::size_t index = 0; index < size; ++index) {
        crc ^= data[index];
        for (int bit = 0; bit < 8; ++bit) {
            const std::uint32_t mask = static_cast<std::uint32_t>(-(crc & 1U));
            crc = (crc >> 1U) ^ (0xEDB88320U & mask);
        }
    }
    return ~crc;
}

std::vector<std::uint8_t> serialize_packet(const Packet& packet) {
    std::vector<std::uint8_t> result(kPacketHeaderSize + packet.payload.size());
    result[0] = 'I';
    result[1] = 'G';
    result[2] = 'N';
    result[3] = 'S';
    result[4] = kProtocolVersion;
    result[5] = static_cast<std::uint8_t>(packet.type);
    write_u16(result.data() + 6, packet.flags);
    write_u32(result.data() + 8, static_cast<std::uint32_t>(packet.payload.size()));
    write_u64(result.data() + 12, packet.sequence);
    write_u64(result.data() + 20, packet.monotonic_ns);
    write_u32(result.data() + 28, crc32(packet.payload.data(), packet.payload.size()));
    std::copy(packet.payload.begin(), packet.payload.end(), result.begin() + kPacketHeaderSize);
    return result;
}

Result<std::vector<Packet>> PacketParser::feed(const std::uint8_t* data, std::size_t size) {
    if (buffer_.size() + size > maximum_payload_ + kPacketHeaderSize)
        return Result<std::vector<Packet>>::failure("connection buffer exceeds maximum packet size");
    buffer_.insert(buffer_.end(), data, data + size);
    std::vector<Packet> packets;
    while (buffer_.size() >= kPacketHeaderSize) {
        const std::uint8_t* header = buffer_.data();
        if (std::memcmp(header, "IGNS", 4) != 0)
            return Result<std::vector<Packet>>::failure("invalid magic");
        if (header[4] != kProtocolVersion)
            return Result<std::vector<Packet>>::failure("unsupported version");
        if (!known_packet_type(header[5]))
            return Result<std::vector<Packet>>::failure("unknown packet type");
        const std::uint32_t payload_length = read_u32(header + 8);
        if (payload_length > maximum_payload_)
            return Result<std::vector<Packet>>::failure("payload exceeds maximum");
        if (buffer_.size() < kPacketHeaderSize + payload_length) break;
        Packet packet;
        packet.type = static_cast<PacketType>(header[5]);
        packet.flags = read_u16(header + 6);
        packet.sequence = read_u64(header + 12);
        packet.monotonic_ns = read_u64(header + 20);
        packet.payload.assign(
            buffer_.begin() + kPacketHeaderSize,
            buffer_.begin() + kPacketHeaderSize + payload_length);
        const std::uint32_t expected_crc = read_u32(header + 28);
        if (crc32(packet.payload.data(), packet.payload.size()) != expected_crc)
            return Result<std::vector<Packet>>::failure("payload CRC mismatch");
        packets.push_back(std::move(packet));
        buffer_.erase(buffer_.begin(), buffer_.begin() + kPacketHeaderSize + payload_length);
    }
    return Result<std::vector<Packet>>::success(std::move(packets));
}

std::vector<std::uint8_t> make_frame_payload(
    const std::string& metadata_json, const std::vector<std::uint8_t>& jpeg) {
    std::vector<std::uint8_t> payload(4 + metadata_json.size() + jpeg.size());
    write_u32(payload.data(), static_cast<std::uint32_t>(metadata_json.size()));
    std::copy(metadata_json.begin(), metadata_json.end(), payload.begin() + 4);
    std::copy(jpeg.begin(), jpeg.end(), payload.begin() + 4 + metadata_json.size());
    return payload;
}

}  // namespace ignis

